File: tests/test_python_compat.rs
use chrono::NaiveDate;
use ibkr_porez::models::*;
use rust_decimal::Decimal;
use std::str::FromStr;
use tempfile::TempDir;

/// Cross-language compatibility test: write files with Rust, verify Python can read them.
#[test]
fn test_python_reads_rust_written_files() {
    let uv = find_uv();

    let temp = TempDir::new().unwrap();
    let data_dir = temp.path().join("data");
    std::fs::create_dir_all(&data_dir).unwrap();
    std::fs::create_dir_all(data_dir.join("declarations")).unwrap();
    std::fs::create_dir_all(data_dir.join("flex-queries")).unwrap();

    write_test_config(&data_dir);
    write_test_transactions(&data_dir);
    write_test_rates(&data_dir);
    write_test_declarations(&data_dir);

    let test_script =
        std::path::Path::new(env!("CARGO_MANIFEST_DIR")).join("tests/test_python_compat.py");

    let output = std::process::Command::new(uv)
        .args([
            "run",
            test_script.to_str().unwrap(),
            data_dir.to_str().unwrap(),
        ])
        .output()
        .expect("Failed to execute `uv run`");

    let stdout = String::from_utf8_lossy(&output.stdout);
    let stderr = String::from_utf8_lossy(&output.stderr);

    assert!(
        output.status.success(),
        "Python compatibility test failed (exit code {:?}):\nstdout:\n{stdout}\nstderr:\n{stderr}",
        output.status.code(),
    );

    assert!(
        stdout.contains("All Python compatibility checks passed"),
        "Unexpected Python output:\n{stdout}",
    );
}

fn find_uv() -> String {
    let output = std::process::Command::new("uv")
        .arg("--version")
        .output()
        .expect("`uv` must be installed to run Python compatibility tests");
    assert!(
        output.status.success(),
        "`uv --version` failed: {}",
        String::from_utf8_lossy(&output.stderr)
    );
    "uv".into()
}

fn write_test_config(data_dir: &std::path::Path) {
    let config = UserConfig {
        full_name: "Test User".into(),
        address: "Test Address".into(),
        data_dir: Some(data_dir.to_string_lossy().into_owned()),
        ..UserConfig::default()
    };
    let json = serde_json::to_string_pretty(&config).unwrap();
    std::fs::write(data_dir.join("config.json"), json).unwrap();
}

fn write_test_transactions(data_dir: &std::path::Path) {
    let txns = vec![Transaction {
        transaction_id: "TEST-001".into(),
        date: NaiveDate::from_ymd_opt(2025, 6, 15).unwrap(),
        r#type: TransactionType::Trade,
        symbol: "AAPL".into(),
        description: "Buy AAPL".into(),
        quantity: Decimal::from_str("10").unwrap(),
        price: Decimal::from_str("150.0").unwrap(),
        amount: Decimal::from_str("-1500.0").unwrap(),
        currency: Currency::USD,
        open_date: None,
        open_price: None,
        exchange_rate: Some(Decimal::from_str("117.25").unwrap()),
        amount_rsd: Some(Decimal::from_str("-175875.0").unwrap()),
    }];
    let json = serde_json::to_string_pretty(&txns).unwrap();
    std::fs::write(data_dir.join("transactions.json"), json).unwrap();
}

fn write_test_rates(data_dir: &std::path::Path) {
    let mut rates = indexmap::IndexMap::new();
    rates.insert("2025-06-15_USD".to_string(), "117.25".to_string());
    let json = serde_json::to_string_pretty(&rates).unwrap();
    std::fs::write(data_dir.join("rates.json"), json).unwrap();
}

fn write_test_declarations(data_dir: &std::path::Path) {
    let decl = Declaration {
        declaration_id: "DECL-001".into(),
        r#type: DeclarationType::Ppdg3r,
        status: DeclarationStatus::Draft,
        period_start: NaiveDate::from_ymd_opt(2025, 1, 1).unwrap(),
        period_end: NaiveDate::from_ymd_opt(2025, 12, 31).unwrap(),
        created_at: chrono::NaiveDateTime::parse_from_str(
            "2025-12-20T10:30:00",
            "%Y-%m-%dT%H:%M:%S",
        )
        .unwrap(),
        submitted_at: None,
        paid_at: None,
        file_path: None,
        xml_content: None,
        report_data: None,
        metadata: indexmap::IndexMap::new(),
        attached_files: indexmap::IndexMap::new(),
    };
    let decl_file = DeclarationsFile {
        declarations: vec![serde_json::to_value(&decl).unwrap()],
        last_declaration_date: Some("2025-12-31".into()),
    };
    let json = serde_json::to_string_pretty(&decl_file).unwrap();
    std::fs::write(data_dir.join("declarations.json"), json).unwrap();
}

File: tests/test_python_compat.py
# /// script
# requires-python = ">=3.12"
# dependencies = ["ibkr-porez @ git+https://github.com/andgineer/ibkr-porez-py.git"]
# ///
"""Cross-language compatibility test: verify Python can read Rust-written JSON files.

Invoked by `uv run` from a Rust integration test:
    uv run tests/test_python_compat.py <data_dir>

Expects transactions.json, rates.json, declarations.json, config.json in <data_dir>.
Uses mock.patch to bypass platformdirs -- no files outside the temp dir.
"""
import json
import sys
from datetime import date
from pathlib import Path
from unittest.mock import patch


def main():
    data_dir = Path(sys.argv[1])
    assert data_dir.exists(), f"data_dir does not exist: {data_dir}"

    from ibkr_porez.models import Currency, UserConfig
    from ibkr_porez.storage import Storage

    # 1. Config: verify Rust-written config.json is readable as UserConfig via Pydantic
    with open(data_dir / "config.json") as f:
        cfg = UserConfig(**json.load(f))
    assert cfg.full_name == "Test User"
    assert cfg.data_dir == str(data_dir)

    # 2. Storage: mock platformdirs so everything stays inside data_dir
    mock_config = UserConfig(
        full_name="Test User",
        address="Test Address",
        data_dir=str(data_dir),
    )
    with (
        patch("ibkr_porez.storage.config_manager") as mock_cm,
        patch("ibkr_porez.storage.user_data_dir", return_value=str(data_dir)),
    ):
        mock_cm.load_config.return_value = mock_config
        storage = Storage()

    # 3. Transactions: pd.read_json must succeed on Rust-written format
    df = storage.get_transactions()
    assert not df.empty, "get_transactions() returned empty DataFrame"
    row = df.iloc[0]
    assert str(row["transaction_id"]) == "TEST-001"
    assert str(row["symbol"]) == "AAPL"
    assert float(row["quantity"]) == 10.0
    assert float(row["amount"]) == -1500.0

    # 4. Exchange rates: json.load + Decimal parsing
    rate = storage.get_exchange_rate(date(2025, 6, 15), Currency.USD)
    assert rate is not None, "get_exchange_rate returned None"
    assert float(rate.rate) == 117.25

    # 5. Declarations: json.load + Declaration(**d) via Pydantic
    decls = storage.get_declarations()
    assert len(decls) >= 1, f"Expected at least 1 declaration, got {len(decls)}"
    decl = decls[0]
    assert decl.declaration_id == "DECL-001"
    assert str(decl.type.value) == "PPDG-3R"

    print("All Python compatibility checks passed.")


if __name__ == "__main__":
    main()
